modules/core/src/main/java/org/apache/ignite/internal/worker/CriticalWorkerWatchdog.java

```diff
@@ -86,8 +90,28 @@ public CompletableFuture<Void> start() {
     }
 
     private void probeLiveness() {
+        Long2LongMap delayedThreadIdsToDelays = getDelayedThreadIdsAndDelays();
+
+        if (delayedThreadIdsToDelays.isEmpty()) {
+            return;
+        }
+
+        ThreadInfo[] threadInfos = threadMXBean.getThreadInfo(delayedThreadIdsToDelays.keySet().toLongArray(), true, true);
+        for (ThreadInfo threadInfo : threadInfos) {
+            if (threadInfo != null) {
+                log.error("A critical thread is blocked for {} ms that is more than the allowed {} ms, it is {}",
+                        delayedThreadIdsToDelays.get(threadInfo.getThreadId()), MAX_ALLOWED_LAG_MS, threadInfo);
+
+                // TODO: IGNITE-16899 - invoke failure handler.
+            }
+        }
+    }
+
+    private Long2LongMap getDelayedThreadIdsAndDelays() {
         long nowNanos = System.nanoTime();
 
+        Long2LongMap delayedThreadIdsToDelays = new Long2LongOpenHashMap();
```

Comment:
I would suggest to use lazy initialization here to avoid spamming unnecessary objects